reactor-netty5-core/src/main/java/reactor/netty5/FutureMono.java

```diff
@@ -0,0 +1,204 @@
+	 *
+	 * @return A {@link Mono} forwarding {@link Future} success or failure
+	 */
+	public static <F extends Future<Void>> Mono<Void> from(F future) {
+		Objects.requireNonNull(future, "future");
+		if (future.isDone()) {
+			if (!future.isSuccess()) {
+				return Mono.error(FutureSubscription.wrapError(future.cause()));
+			}
+			return Mono.empty();
+		}
+		return new ImmediateFutureMono<>(future);
+	}
+
+	/**
+	 * Convert a supplied {@link Future} for each subscriber into {@link Mono}.
+	 * {@link Mono#subscribe(Subscriber)}
+	 * will bridge to {@link Future#addListener(FutureListener)}.
+	 *
+	 * @param deferredFuture the future to evaluate and convert from
+	 * @param <F> the future type
+	 *
+	 * @return A {@link Mono} forwarding {@link Future} success or failure
+	 */
+	public static <F extends Future<Void>> Mono<Void> deferFuture(Supplier<F> deferredFuture) {
+		return new DeferredFutureMono<>(deferredFuture);
+	}
+
+	final static class ImmediateFutureMono<F extends Future<Void>> extends FutureMono {
+
+		final F future;
+
+		ImmediateFutureMono(F future) {
+			this.future = Objects.requireNonNull(future, "future");
+		}
+
+		@Override
+		public void subscribe(final CoreSubscriber<? super Void> s) {
+			doSubscribe(s, future);
+		}
+	}
+
+	final static class DeferredFutureMono<F extends Future<Void>> extends FutureMono {
+
+		final Supplier<F> deferredFuture;
+
+		DeferredFutureMono(Supplier<F> deferredFuture) {
+			this.deferredFuture =
+					Objects.requireNonNull(deferredFuture, "deferredFuture");
+		}
+
+		@Override
+		public void subscribe(CoreSubscriber<? super Void> s) {
+			F f;
+			try {
+				f = deferredFuture.get();
+			}
+			catch (Throwable t) {
+				Operators.error(s, t);
+				return;
+			}
+
+			if (f == null) {
+				Operators.error(s,
+						Operators.onOperatorError(new NullPointerException(
+								"Deferred supplied null"), s.currentContext()));
+				return;
+			}
+
+			doSubscribe(s, f);
+		}
+	}
+
+	static <F extends Future<Void>> void doSubscribe(CoreSubscriber<? super Void> s, F future) {
+		if (future.isDone()) {
+			if (future.isSuccess()) {
+				Operators.complete(s);
+			}
+			else {
+				Operators.error(s, FutureSubscription.wrapError(future.cause()));
+			}
+			return;
+		}
+
+		FutureSubscription<F> fs = new FutureSubscription<>(future, s);
+		// propagate subscription before adding listener to avoid any race between finishing future and onSubscribe
+		// is called
+		s.onSubscribe(fs);
+
+		// check if subscription was not cancelled immediately.
+		if (fs.cancelled) {
+			// if so do nothing anymore
+			return;
+		}
+
+		// add listener to the future to propagate on complete when future is done
+		// addListener likely to be thread safe method
+		future.addListener(fs);
+
+		// check once again if is cancelled to see if we need to removeListener in case addListener racing with
+		// subscription.cancel (which should remove listener)
+		if (fs.cancelled) {
+			// future.removeListener(fs); TODO
+		}
```

Comment:
Did you leave the line 147 commented on purpose ?
I ask this because apparently, listeners can't be removed anymore from Future in netty5. So is there something else that still needs to be implemented in case the subscription is cancelled ? If not, then can we remove the lines between L144-L148 ?

Reply:
For the moment we will keep these TODOs until we are sure that we don't need them. This will happen post M3 release.